### DISCUSSION TITLE
CS1540 - accessing base class protected member from child class's nested class

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6598] [General]
The following code: [SharpLab](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE8QBwCcB7AFwFMBjcgE0QEsA7ExAfQA8BubXQ0ymxAEMARgGcSBQVWRhEAIUGiy3HFhi4N+YuSplaCpQAp0RAO4MyBAJQrNazQ8QkAFnVEA6FmYsFEAXkRvSy51RwBfWFsNBGQAFkQAMSIiQytMJ1cPL3NLTzZ/RAgORAiYKJ5tfj1EAHoAKkZyAgZBABs6msQCMkFqIgZWgE80ViCCEI1S0piUABEALlRYe1wQeKSUtIwMtzyCopLbFFRXVtp5+UVlUNVynBiTujPDWcCc60QL4SvDMbS79IAjQ1GouXbZHwFMYTcIAmbxBQEVLpMFZMZ7AIHUqOWo1Xg6ATdXr9IaIV4Q4K2KZAA==)
```cs
public class C {
    protected int _x;
    protected abstract class Base
    {
        protected Base(C owner)
        {
            this._owner = owner;
        }

        public void Foo() { this._owner._x = 1; }

        protected /*internal*/ readonly C _owner;
    }
}
public class D : C
{
    void Foo() { this._x = 1; }
    class Child : Base
    {
        public Child(D owner) : base(owner) 
        { 
            //this._owner = owner;
        }
        public void Bar() { this._owner._x = 1; }
        //protected readonly D _owner;
    }
}
```
Produces `error CS1540: Cannot access protected member 'C._x' via a qualifier of type 'C'; the qualifier must be of type 'D.Child' (or derived from it)`
Shouldn't this be legal? After all, `Child` is nested in `D`, and `D` is a subclass of `C`, so `C` protected members are visible to `D` and by extension to everything nested in `D`. Am I missing anything?

**Comment by bernd5** [comment]
The wording is:

> A protected member is accessible within its class and by derived class instances

the problem here is, that C could be something else than a C or D it could be E which inherits from C, too.
Therefore the rule is violated (it is not *within its class* and not by *by derived class instance*).

You could change `Bar()` to:
```csharp
public void Bar()
{ 
    if (this._owner is D d)
    {
        d._x = 1; 
    }
}
```

Or you could make `Child` a nested class of `C` - then `C` and `Child` would become Friends which allows unrestricted access.
